### PR TITLE
Add Amazon DynamoDB SDK2 extension

### PR DESCRIPTION
**Pull request by marcinczeczko**
Please find at cleaned version of #2050 

An extension that provides AWS DynamoDB client based on AWS SDK2 that supports async netty client.
An extension adds support for both Async (Netty) and Blocking (Apache) HTTP clients.

### CODE REVIEW

**Comment by marcinczeczko** [comment]
@gsmet - please fine the PR checks for the issue I mentioned in #2050. Basically only "Build_JDK11_Linux" pipeline failed because the local dynamoDB for integration testing (that's started asynchronously via maven-exec-plugin) is getting up after tests started, as shown in azure pipline:
```
INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running io.quarkus.it.dynamodb.DynamoDbFunctionalityTest
2019-04-23 21:02:00,612 INFO  [io.qua.dep.QuarkusAugmentor] (main) Beginning quarkus augmentation
2019-04-23 21:02:02,654 INFO  [io.qua.dep.QuarkusAugmentor] (main) Quarkus augmentation completed in 2042ms
2019-04-23 21:02:03,608 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT started in 0.914s. Listening on: http://[::]:8081
2019-04-23 21:02:03,609 INFO  [io.quarkus] (main) Installed features: [cdi, dynamodb-sdk2, resteasy]
2019-04-23 21:02:04,647 INFO  [io.qua.it.dyn.DynamoDBResource] (executor-thread-1) Testing Async with table: async
Initializing DynamoDB Local with the following configuration:
Port:	41769
InMemory:	true
DbPath:	null
SharedDb:	false
shouldDelayTransientStatuses:	false
CorsParams:	*

2019-04-23 21:02:06,853 ERROR [org.jbo.res.res.i18n] (aws-java-sdk-NettyEventLoop-1-3) RESTEASY002025: Unknown exception while executing GET /async: java.util.concurrent.CompletionException: software.amazon.awssdk.core.exception.SdkClientException
	at software.amazon.awssdk.utils.CompletableFutureUtils.errorAsCompletionException(CompletableFutureUtils.java:61)
```

However, Windows or Build_Native_Linux pipelines are just fine. For instance how it looks like on windows one in the build log:
```
[INFO] --- maven-surefire-plugin:2.22.0:test (default-test) @ quarkus-integration-test-aws-dynamodb-sdk2 ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
Initializing DynamoDB Local with the following configuration:
Port:	2915
InMemory:	true
DbPath:	null
SharedDb:	false
shouldDelayTransientStatuses:	false
CorsParams:	*

[INFO] Running io.quarkus.it.dynamodb.DynamoDbFunctionalityTest
2019-04-23 20:52:10.117:INFO::main: Logging initialized @3216ms to org.eclipse.jetty.util.log.StdErrLog
2019-04-23 20:52:10,746 INFO  [io.qua.dep.QuarkusAugmentor] (main) Beginning quarkus augmentation
2019-04-23 20:52:13,451 INFO  [io.qua.dep.QuarkusAugmentor] (main) Quarkus augmentation completed in 2705ms
2019-04-23 20:52:14,732 INFO  [io.quarkus] (main) Quarkus 999-SNAPSHOT started in 1.237s. Listening on: http://[::]:8081
2019-04-23 20:52:14,732 INFO  [io.quarkus] (main) Installed features: [cdi, dynamodb-sdk2, resteasy]
2019-04-23 20:52:16,078 INFO  [io.qua.it.dyn.DynamoDBResource] (executor-thread-1) Testing Async with table: async
2019-04-23 20:52:17,747 INFO  [io.qua.it.dyn.DynamoDBResource] (executor-thread-1) Testing Blocking with table: blocking
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 9.656 s - in io.quarkus.it.dynamodb.DynamoDbFunctionalityTest
```

I experimented with various phases for exec plugin but I'm stuck.

**Comment by marcinczeczko** [comment]
@gsmet - I managed to fix the issue. DynamoDB local is started as QuarkusTestResource instead of via maven exec plugin, so now it's ready for review if you're interested in that extension

File: test-framework/amazon-dynamodb/pom.xml
Comment (gsmet) [inline test-framework/amazon-dynamodb/pom.xml]: What is this repository? We clearly can't add an external repository like that.

**Review by gsmet — CHANGES_REQUESTED** [review]
Definitely not a complete review but there's one thing bothering me in the test pom. Could you take a look?

File: test-framework/amazon-dynamodb/pom.xml
Comment (marcinczeczko) [reply test-framework/amazon-dynamodb/pom.xml]: You're right. It shouldn't be here, I must forgot to remove it. Already fixed.

**Comment by marcinczeczko** [comment]
@gsmet - unnecessary repository definition removed from pom.xml

**Comment by gsmet** [comment]
@marcinczeczko sorry for the delay. Apparently, there are CI failures but the build information is not accessible anymore...

Could you fix the conflicts and push again so that I can see what's going on?

**Comment by marcinczeczko** [comment]
@gsmet I fixed conflicts. Build info is not available probably because a PR is blocked on changes you requested.

**Comment by marcinczeczko** [comment]
@gsmet apparently a DynamoDBLocal artifact is not available at maven central. I had to add a dedicated repository to the azure pipelines settings.xml

**Comment by gsmet** [comment]
@marcinczeczko that's not something we can accept unfortunately. We need it to be available on Central. Could you reach out to the developers of this artifact and see if they can push it?

**Comment by marcinczeczko** [comment]
@gsmet I'll reach up AWS support (as it's provided by them) if there is any chance for that. If it will not be possible an alternate solution is to use dynamodb local as docker image, run similary as for the mysql integration tests. Is it an acceptable solution ?

**Comment by gsmet** [comment]
Yes, using Docker is OK as long as the tests are not included in the default build. Just mimic what we did with the MariaDB/PostgreSQL integration tests.

**Comment by drissamri** [comment]
Great work @marcinczeczko, hope to see this finalized & merged soon!

**Comment by marcinczeczko** [comment]
@gsmet - I changed the integration tests to use dynamodb local as docker container.

**Comment by gsmet** [comment]
Hmmm, CI is failing with:
```
Unable to find image 'dynamodb:latest' locally
docker: Error response from daemon: pull access denied for dynamodb, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
See 'docker run --help'.
```

**Comment by marcinczeczko** [comment]
Looks like I forgot to setup proper image name. Already fixed

**Comment by gsmet** [comment]
@marcinczeczko I'll make a proper review next week.

**Comment by drissamri** [comment]
Any idea when this can be merged and when the next release with it will be? Curious to try this out

**Comment by gsmet** [comment]
I will review it on Monday, I was at a conference last week. Then probably a few iterations so it might not make next Wednesday's release but it should make it for the next one.

**Review by gsmet — COMMENTED** [review]
I added some comments and questions inline.

File: bom/pom.xml
Comment (marcinczeczko) [reply bom/pom.xml]: Good point, it's released together. I'll fix that, and rename it to simply to `awssdk.version`

File: bom/pom.xml
Comment (marcinczeczko) [reply bom/pom.xml]: It's a valid point. I'd rather remove `-sdk2`. Personally, I don't think it's worth to look back at 1.x sdk as it doesn't support async http, nor it's so extensible as 2.x

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java
Comment (marcinczeczko) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java]: Same as above. I'll remove `-sdk2`

File: integration-tests/aws-dynamodb-sdk2/src/main/java/io/quarkus/it/dynamodb/DynamoDBResource.java
Comment (marcinczeczko) [reply integration-tests/aws-dynamodb-sdk2/src/main/java/io/quarkus/it/dynamodb/DynamoDBResource.java]: I think it sounds like a good plan :)
Btw, there might be a plenty of such knobs, as the SDK2 clients are nicely configurable via builders for various aspects, e.g. builders for http clients (apache or netty based) to configure http transport stuff, etc.
Anyway, I need to extract possible config options that make sense, and propose a solution.

File: extensions/amazon-dynamodb-sdk2/deployment/src/main/java/io/quarkus/dynamodbsdk2/deployment/DynamodbSdk2Processor.java
Comment (marcinczeczko) [reply extensions/amazon-dynamodb-sdk2/deployment/src/main/java/io/quarkus/dynamodbsdk2/deployment/DynamodbSdk2Processor.java]: Thanks, will remove it

File: bom/pom.xml
Comment (gsmet) [reply bom/pom.xml]: OK, let's do that then (and let's be consistent and remove it everywhere).

File: integration-tests/aws-dynamodb-sdk2/src/main/java/io/quarkus/it/dynamodb/DynamoDBResource.java
Comment (gsmet) [reply integration-tests/aws-dynamodb-sdk2/src/main/java/io/quarkus/it/dynamodb/DynamoDBResource.java]: Sounds like a plan. The nice thing is that we can produce application scoped beans: if they are not used, they will be removed.

This way, one can create a specifically configured object if it needs some more fine grained configuration.

File: extensions/amazon-dynamodb-sdk2/deployment/src/main/java/io/quarkus/dynamodbsdk2/deployment/DynamodbSdk2Processor.java
Comment (marcinczeczko) [reply extensions/amazon-dynamodb-sdk2/deployment/src/main/java/io/quarkus/dynamodbsdk2/deployment/DynamodbSdk2Processor.java]: I'll check this. But as far as I remember the SDK uses slf4j so should be fine. I'll verify it after addressing the comment for bean producer.

File: bom/pom.xml
Comment (marcinczeczko) [reply bom/pom.xml]: Done

File: core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java
Comment (marcinczeczko) [reply core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java]: Done

File: extensions/amazon-dynamodb-sdk2/deployment/src/main/java/io/quarkus/dynamodbsdk2/deployment/DynamodbSdk2Processor.java
Comment (marcinczeczko) [reply extensions/amazon-dynamodb-sdk2/deployment/src/main/java/io/quarkus/dynamodbsdk2/deployment/DynamodbSdk2Processor.java]: Done

**Comment by gsmet** [comment]
@marcinczeczko do you plan to work more on this one soon?

I think it's a valuable addition anyway so maybe we should merge it and iterate later.

Let me know what you think about it.

In any case, could you squash everything and force-push?

Thanks!

**Comment by marcinczeczko** [comment]
Hi @gsmet sorry for not responding. I was off for some time so didn't get any progress as for now. I plan to come back to that improvements next month when come back. However, you're proposal is fine to me, we can progress with iterative approach. I'll update current PR with squashed commit then.

**Comment by marcinczeczko** [comment]
@gsmet - squashed commits pushed

**Comment by dimidevk** [comment]
is there something, that has to be done, in order to merge this? would be really great to see this feature in the next release..

File: extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java
Comment (gsmet) [inline extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java]: It would be better to use the JBoss Logging wrappers instead. Could you check if it's possible (basically, you would have to exclude the dependencies and use JBoss Logging dependencies instead, we have one for commons-logging and one for the JDK logging if need be)?

File: extensions/amazon-dynamodb/runtime/pom.xml
Comment (gsmet) [inline extensions/amazon-dynamodb/runtime/pom.xml]: Why this dependency?

File: extensions/amazon-dynamodb/runtime/pom.xml
Comment (gsmet) [inline extensions/amazon-dynamodb/runtime/pom.xml]: Why this dependency?

**Review by gsmet — COMMENTED** [review]
@marcinczeczko I added 3 more comments:

The one about JBoss Logging, I think we can fix later but it would be nice if we could converge of the JSON dependencies before tomorrow so that I can include this one in the next release.

Thanks!

File: extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java
Comment (marcinczeczko) [reply extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java]: I wasn't able to find any existing extension that swtichers to jboss logging. Any extension that registers Jdk14Logger & LogFactoryImpl for reflections doesn't have any modifications to the dependencies.  Did your intention is to simply exclude commons logging dependency and include jboss one ?

File: extensions/amazon-dynamodb/runtime/pom.xml
Comment (marcinczeczko) [reply extensions/amazon-dynamodb/runtime/pom.xml]: removed

File: extensions/amazon-dynamodb/runtime/pom.xml
Comment (marcinczeczko) [reply extensions/amazon-dynamodb/runtime/pom.xml]: removed

File: extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java
Comment (marcinczeczko) [reply extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java]: I configured DEBUG logging of the dynamodb clients and looks like the underlying logging from quarkus was used. Added application properties to enable it

Comment (gsmet) [reply extensions/amazon-dynamodb/deployment/src/main/java/io/quarkus/dynamodb/deployment/DynamodbProcessor.java]: Yes, the idea is to exclude the dependencies and then you shouldn't need to register these classes for reflection. At least if it works correctly :).

**Comment by marcinczeczko** [comment]
@gsmet - I did following
```
        <dependency>
            <groupId>software.amazon.awssdk</groupId>
            <artifactId>apache-client</artifactId>
            <exclusions>
                <exclusion>
                    <groupId>commons-logging</groupId>
                    <artifactId>commons-logging</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>org.jboss.logging</groupId>
            <artifactId>jboss-logging</artifactId>
        </dependency>
```
But unfortunately, native image building fails
```
Error: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved type during parsing: org.apache.commons.logging.LogFactory. To diagnose the issue you can use the --allow-incomplete-classpath option. The missing type is then reported at run time when it is accessed the first time.
Trace: 
        at parsing org.apache.http.impl.client.BasicAuthCache.<init>(BasicAuthCache.java:63)
Call path from entry point to org.apache.http.impl.client.BasicAuthCache.<init>(SchemePortResolver): 
```
So looks like it would require more work.

**Comment by gsmet** [comment]
@marcinczeczko you would need this artifact too AFAIK: https://search.maven.org/artifact/org.jboss.logging/commons-logging-jboss-logging/1.0.0.Final/jar

**Comment by marcinczeczko** [comment]
Yep, in the mean time I found the same
```
        <dependency>
            <groupId>software.amazon.awssdk</groupId>
            <artifactId>apache-client</artifactId>
            <exclusions>
                <exclusion>
                    <groupId>commons-logging</groupId>
                    <artifactId>commons-logging</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>org.jboss.logging</groupId>
            <artifactId>jboss-logging</artifactId>
        </dependency>
        <dependency>
            <groupId>org.jboss.logmanager</groupId>
            <artifactId>commons-logging-jboss-logmanager</artifactId>
            <version>1.0.0.Final</version>
        </dependency>
```
But native build fails too, with same errors
```
Error: com.oracle.graal.pointsto.constraints.UnresolvedElementException: Discovered unresolved type during parsing: org.apache.commons.logging.LogFactory. To diagnose the issue you can use the --allow-incomplete-classpath option. The missing type is then reported at run time when it is accessed the first time.
Trace: 
        at parsing org.apache.http.impl.execchain.RetryExec.<init>(RetryExec.java:63)
Call path from entry point to org.apache.http.impl.execchain.RetryExec.<init>(ClientExecChain, HttpRequestRetryHandler): 
        at org.apache.http.impl.execchain.RetryExec.<init>(RetryExec.java:70)
```

**Comment by gsmet** [comment]
@marcinczeczko that's not the artifact you're looking for :).

See my earlier comment for the right one.

**Comment by marcinczeczko** [comment]
Sorry, I missed the proper groupId. Unfortunately, native build fails exactly the same way.

**Comment by marcinczeczko** [comment]
@gsmet  - maybe it would require graalvm substitutions in the https://github.com/quarkusio/quarkus/blob/master/core/runtime/src/main/java/io/quarkus/runtime/graal/LoggingSubstitutions.java to cover commons logging

**Comment by gsmet** [comment]
I don't think so. It complains about the class being absent.

The class is here: https://github.com/jboss-logging/commons-logging-jboss-logging/tree/master/src/main/java/org/apache/commons/logging .

Are you sure you're pointing to the artifact I mentioned (the artifactId was incorrect too) and the dependency is in the runtime artifact (not the deployment one)?

**Comment by marcinczeczko** [comment]
Ohh my, looks like I did two mistakes on two lines :) Yes, it's in runtime module, and now it looks like below
```
        <dependency>
            <groupId>software.amazon.awssdk</groupId>
            <artifactId>apache-client</artifactId>
            <exclusions>
                <exclusion>
                    <groupId>commons-logging</groupId>
                    <artifactId>commons-logging</artifactId>
                </exclusion>
            </exclusions>
        </dependency>
        <dependency>
            <groupId>org.jboss.logging</groupId>
            <artifactId>commons-logging-jboss-logging</artifactId>
            <version>1.0.0.Final</version>
        </dependency>
```

And of course native build finishes with success  :)

Do you want to add this dependency to the bom ?

**Comment by marcinczeczko** [comment]
Anyway, see the changes now.

**Comment by gsmet** [comment]
@marcinczeczko yes, we need to add it to the BOM with the others. Preferably near the other JBoss Logging dependencies.

**Comment by marcinczeczko** [comment]
@gsmet done. Build finished with success

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
@marcinczeczko thanks for your work.

I will merge this one and we can refine from there.

**Comment by drissamri** [comment]
@marcinczeczko I'm trying to have a go with this, I'm not that familiar with Quarkus yet so I guess I can't just initialize the bean myself? 

```
 private final DynamoDbClient dynamoDbClient;

    public ProductService() {
        final DynamoDbClient dynamoDbClient = DynamoDbClient.builder()
                .region(Region.EU_WEST_1)
                .build();
        this.dynamoDbClient = dynamoDbClient;

    }
```
```
Error: No instances are allowed in the image heap for a class that is initialized or reinitialized at image runtime: sun.security.provider.NativePRNG. Try marking this class for build-time initialization with --initialize-at-build-time=sun.security.provider.NativePRNG
Trace: 	object java.security.SecureRandom
	object sun.security.ssl.SSLContextImpl$TLSContext
	object sun.security.ssl.SSLSocketFactoryImpl
	object software.amazon.awssdk.http.apache.internal.conn.SdkTlsSocketFactory
	object java.util.concurrent.ConcurrentHashMap$Node
	object java.util.concurrent.ConcurrentHashMap$Node[]
	object java.util.concurrent.ConcurrentHashMap
	object org.apache.http.config.Registry
	object org.apache.http.impl.conn.DefaultHttpClientConnectionOperator
	object org.apache.http.impl.conn.PoolingHttpClientConnectionManager
	object software.amazon.awssdk.http.apache.internal.conn.ClientConnectionManagerFactory$Handler
	object com.sun.proxy.$Proxy186
	object org.apache.http.impl.client.HttpClientBuilder$2
	object java.lang.Object[]
	object java.util.ArrayList
	object org.apache.http.impl.client.InternalHttpClient
	object software.amazon.awssdk.http.apache.internal.impl.ApacheSdkHttpClient
	object software.amazon.awssdk.http.apache.ApacheHttpClient
	object java.util.HashMap$Node
	object java.util.HashMap$Node
	object java.util.HashMap$Node[]
	object java.util.HashMap
	object software.amazon.awssdk.utils.AttributeMap
	object software.amazon.awssdk.core.client.config.SdkClientConfiguration
	object software.amazon.awssdk.services.dynamodb.DefaultDynamoDbClient
	object be.drissamri.ProductService_ClientProxy
	object be.drissamri.ProductService_Bean
	object java.lang.Object[]
	object java.util.ArrayList
	object io.quarkus.arc.ArcContainerImpl
	object java.util.concurrent.atomic.AtomicReference
	method io.quarkus.arc.Arc.container()
Call path from entry point to io.quarkus.arc.Arc.container():
	at io.quarkus.arc.Arc.container(Arc.java:30)
	at be.drissamri.ProductService_ClientProxy.arc$delegate(ProductService_ClientProxy.zig:67)
	at be.drissamri.ProductService_ClientProxy.addProduct(ProductService_ClientProxy.zig:158)
	at be.drissamri.ProductLambda.handleRequest(ProductLambda.java:30)
	at be.drissamri.ProductLambda.handleRequest(ProductLambda.java:16)
	at io.quarkus.amazon.lambda.runtime.AmazonLambdaRecorder$2.run(AmazonLambdaRecorder.java:65)
	at java.lang.Thread.run(Thread.java:748)
	at com.oracle.svm.core.thread.JavaThreads.threadStartRoutine(JavaThreads.java:473)
	at com.oracle.svm.core.posix.thread.PosixJavaThreads.pthreadStartRoutine(PosixJavaThreads.java:193)
	at com.oracle.svm.core.code.IsolateEnterStub.PosixJavaThreads_pthreadStartRoutine_e1f4a8c0039f8337338252cd8734f63a79b5e3df(generated:0)
```

```
Call path from entry point to software.amazon.awssdk.http.apache.internal.conn.IdleConnectionReaper.getInstance():
```

**Comment by bnusunny** [comment]
@drissamri I created one working example. You can find it [here](https://github.com/bnusunny/supersonic-java-on-lambda/tree/master/quarkus-on-lambda-dynamodb).

**Comment by dimidevk** [comment]
@drissamri here is another [one](https://github.com/devk-insurance/quarkus-amazon-lambda), very similar to the one from @bnusunny :smile:

**Comment by gsmet** [comment]
@drissamri could you open an issue with some sort of reproducer? It's not a good idea to start a discussion in a merged PR. Thanks!

**Pull request merged**
